src/epic/distances/epic.py

```diff
@@ -0,0 +1,176 @@
+            """Canonical reward function.
+            ``state``, ``action``, ``next_state`` correspond to s, a, s' in the paper
+            and ``state_sample``, ``action_sample``, ``next_state_sample``
+            correspond to S, A, S' in the paper.
+
+            Args:
+                state: The batch of state samples from the coverage distribution.
+                action: The batch of action samples from the coverage distribution.
+                next_state: The batch of next state samples from the coverage distribution.
+                done: The batch of done samples from the coverage distribution.
+
+            Returns:
+                The canonicalized reward function.
+            """
+            n_samples_cov = state.shape[0]
+
+            assert n_samples_cov == action.shape[0] == next_state.shape[0] == done.shape[0]
+
+            n_samples = n_samples_cov * n_samples_can
+
+            # Copy each sample in n_samples_cov to n_samples_can times.
+            state, action, next_state, done = utils.broadcast(state, action, next_state, done, n_samples_can)
+
+            # Create n_samples_cov * n_samples_can samples.
+            _, state_sample = self.state_sampler.sample(n_samples)
+            action_sample = self.action_sampler.sample(n_samples)
+            done_sample, next_state_sample = self.state_sampler.sample(n_samples)
+
+            # Reshape to (n_samples_cov, n_samples_can, -1).
+            # This is easier than reshaping the output of each flat reward function call
+            # to take the mean along the inner monte carlo estimator.
+            state_sample, action_sample, next_state_sample, done_sample = utils.reshape(
+                state_sample, action_sample, next_state_sample, done_sample, n_samples_cov
+            )
+
+            # E[R(s', A, S')]. We sample action and next state,
+            # and pass in ``next_state`` as the state.
+            # We make this the first dimension to then take the mean.
+            term_1 = np.mean(rew_fn(action_sample, next_state, next_state_sample, done_sample, batch_dims=2), axis=0)
+            # E[R(s, A, S')]. We also sample action and next state.
+            # Now it's simply ``state`` that we pass in as the state.
+            term_2 = np.mean(rew_fn(action_sample, state, next_state_sample, done_sample, batch_dims=2), axis=0)
+            # E[R(S, A, S')]. We sample state, action, and next state.
+            # This does not require cartesian product over batches
+            # as it's not a random variable.
+            term_3 = np.mean(rew_fn(action_sample, state_sample, next_state_sample, done_sample), axis=0)
+
+            return (
+                reward_function(state, action, next_state, done)
+                + self.discount_factor * term_1
+                - term_2
+                - self.discount_factor * term_3
+            )
+
+        return canonical_reward_fn
+
+    def _distance(
```

Comment:
divergence-free rewards will use the same `_distance` implementation, so we'll need to move this somewhere reusable

Reply:
I'd be in favor of a `PearsonDistanceMixin` that both `EPIC` and `DivergenceFree` inherit from.

Reply:
Sounds like a good solution to me!